cpp/src/arrow/dataset/dataset_encryption_test.cc

```diff
@@ -0,0 +1,388 @@
+      encryption_config->footer_key = footer_key_name;
+    }
+
+    // DatasetEncryptionConfiguration
+    DatasetEncryptionConfiguration dataset_encryption_config;
+    dataset_encryption_config.crypto_factory = crypto_factory;
+    dataset_encryption_config.encryption_config = encryption_config;
+
+    // DatasetDecryptionConfiguration
+    DatasetDecryptionConfiguration dataset_decryption_config;
+    dataset_decryption_config.crypto_factory = crypto_factory;
+
+    // create our Parquet file format object
+    auto file_format = std::make_shared<ParquetFileFormat>();
+
+    file_format->SetDatasetEncryptionConfig(
+        std::make_shared<DatasetEncryptionConfiguration>(dataset_encryption_config));
+    file_format->SetDatasetDecryptionConfig(
+        std::make_shared<DatasetDecryptionConfiguration>(dataset_decryption_config));
+
+    return file_format;
+  }
+
+  // utility to build the key map
+  std::unordered_map<std::string, std::string> BuildKeyMap(
+      const std::string_view* column_ids, const std::string_view* column_keys,
+      int num_columns, const std::string_view& footer_id,
+      const std::string_view& footer_key) {
+    std::unordered_map<std::string, std::string> key_map;
+    // add column keys
+    for (int i = 0; i < num_columns; i++) {
+      key_map.insert({std::string(column_ids[i]), std::string(column_keys[i])});
+    }
+    // add footer key
+    key_map.insert({std::string(footer_id), std::string(footer_key)});
+
+    return key_map;
+  }
+
+  // A utility function to validate our files were written out */
+  void ValidateFilesExist(const std::shared_ptr<arrow::fs::internal::MockFileSystem>& fs,
+                          const std::vector<std::string>& files) {
+    for (const auto& file_path : files) {
+      ASSERT_OK_AND_ASSIGN(auto result, fs->GetFileInfo(file_path));
+
+      ASSERT_NE(result.type(), arrow::fs::FileType::NotFound);
+    }
+  }
+
+  // Build a dummy table
+  std::shared_ptr<::arrow::dataset::InMemoryDataset> BuildTable() {
+    // Create an Arrow Table
+    auto schema = arrow::schema(
+        {arrow::field("a", arrow::int64()), arrow::field("b", arrow::int64()),
+         arrow::field("c", arrow::int64()), arrow::field("part", arrow::utf8())});
+
+    std::vector<std::shared_ptr<arrow::Array>> arrays(4);
+    arrow::NumericBuilder<arrow::Int64Type> builder;
+    ARROW_EXPECT_OK(builder.AppendValues({0, 1, 2, 3, 4, 5, 6, 7, 8, 9}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[0]));
+    builder.Reset();
+    ARROW_EXPECT_OK(builder.AppendValues({9, 8, 7, 6, 5, 4, 3, 2, 1, 0}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[1]));
+    builder.Reset();
+    ARROW_EXPECT_OK(builder.AppendValues({1, 2, 1, 2, 1, 2, 1, 2, 1, 2}));
+    ARROW_EXPECT_OK(builder.Finish(&arrays[2]));
+    arrow::StringBuilder string_builder;
+    ARROW_EXPECT_OK(
+        string_builder.AppendValues({"a", "b", "c", "d", "e", "f", "g", "h", "i", "j"}));
+    ARROW_EXPECT_OK(string_builder.Finish(&arrays[3]));
```

Comment:
You may actually call `TableFromJSON` to create a table directly from a json string.